commons-mongo/src/main/scala/com/avsystem/commons/mongo/typed/TypedMongoCollection.scala

```diff
@@ -0,0 +1,292 @@
+
+    Observable.fromReactivePublisher(publisherWithOptions).map(property.format.readBson)
+  }
+
+  def insertOne(
+    value: E,
+    setupOptions: InsertOneOptions => InsertOneOptions = identity
+  ): Task[Unit] =
+    single(nativeCollection.insertOne(value, setupOptions(new InsertOneOptions))).void
+
+  def insertMany(
+    values: Seq[E],
+    setupOptions: InsertManyOptions => InsertManyOptions = identity
+  ): Task[Unit] =
+    single(nativeCollection.insertMany(values.asJava, setupOptions(new InsertManyOptions))).void
+
+  def deleteOne(
+    filter: MongoDocumentFilter[E],
+    setupOptions: DeleteOptions => DeleteOptions = identity
+  ): Task[DeleteResult] =
+    single(nativeCollection.deleteOne(filter.toBson, setupOptions(new DeleteOptions)))
+
+  def deleteMany(
+    filter: MongoDocumentFilter[E],
+    setupOptions: DeleteOptions => DeleteOptions = identity
+  ): Task[DeleteResult] =
+    single(nativeCollection.deleteMany(filter.toBson, setupOptions(new DeleteOptions)))
+
+  def updateOne(
+    filter: MongoDocumentFilter[E],
+    update: MongoDocumentUpdate[E],
+    upsert: Boolean = false, // extracted as separate param because it's very commonly used
+    setupOptions: UpdateOptions => UpdateOptions = identity
+  ): Task[UpdateResult] = {
+    val options = setupOptions(new UpdateOptions).upsert(upsert)
+    val (updateBson, arrayFilters) = update.toBsonAndArrayFilters
+    if (!arrayFilters.isEmpty) {
+      options.arrayFilters(arrayFilters)
+    }
+    single(nativeCollection.updateOne(filter.toBson, updateBson, options))
+  }
+
+  def updateMany(
+    filter: MongoDocumentFilter[E],
+    update: MongoDocumentUpdate[E],
+    upsert: Boolean = false, // extracted as separate param because it's very commonly used
+    setupOptions: UpdateOptions => UpdateOptions = identity
+  ): Task[UpdateResult] = {
+    val options = setupOptions(new UpdateOptions).upsert(upsert)
+    val (updateBson, arrayFilters) = update.toBsonAndArrayFilters
+    if (!arrayFilters.isEmpty) {
+      options.arrayFilters(arrayFilters)
+    }
+    single(nativeCollection.updateMany(filter.toBson, updateBson, options))
+  }
+
+  def replaceOne(
+    filter: MongoDocumentFilter[E],
+    replacement: E,
+    upsert: Boolean = false, // extracted as separate param because it's very commonly used
+    setupOptions: ReplaceOptions => ReplaceOptions = identity
+  ): Task[UpdateResult] = {
+    val options = setupOptions(new ReplaceOptions).upsert(upsert)
+    single(nativeCollection.replaceOne(filter.toBson, replacement, options))
+  }
+
+  def bulkWrite(
+    writes: Seq[MongoWrite[E]],
+    setupOptions: BulkWriteOptions => BulkWriteOptions = identity
+  ): Task[BulkWriteResult] = {
+    val requests = writes.iterator.map(_.toWriteModel).to[JList]
+    single(nativeCollection.bulkWrite(requests, setupOptions(new BulkWriteOptions)))
+  }
+
+  def createIndex(
+    index: MongoIndex[E],
+    setupOptions: IndexOptions => IndexOptions = identity
```

Comment:
If we can afford to be opinionated, please add `background(true)`

Reply:
Same remark as before. This would break the assumption that `TypedMongoCollection` is just a better-typed wrapper over the native collection, without changing semantics.